### PR TITLE
refactor(pool-header): retire Health Score; move breaches to a tab with chart

### PR DESCRIPTION
**Pull request by chapati23**
## Why

The pool-detail page had two near-identical health-% tiles drifting apart visually:

- **Health Score**: "% of time strictly under threshold" (pre-rule-change metric; 0-second grace)
- **Uptime**: "% of time NOT in CRITICAL state" (current metric; 1h grace)

Under the new health rule shipped in PR #193/#194, short breaches aren't warnings, so Health Score's numerator became meaningless. The tiles were showing wildly different numbers (0.00% vs 4.981% on some pools) and confusing users. Keeping both also meant the dashboard was implicitly asking the viewer to pick which rule they trusted.

## What

**Removed** — `HealthScoreValue` + `HealthScoreInfoIcon`, `useHealthScore` hook, `pool-health-score.ts` helpers (`formatBinaryHealthPct`, `computeBinaryHealthWindow`), their test suites, and the `healthBinarySeconds` / `healthBinaryValue` / `lastOracleSnapshotTimestamp` / `lastDeviationRatio` fields from the UI types and queries.

**Kept** — `healthTotalSeconds` on `Pool` (Uptime's denominator). Indexer-side fields stay populated; no schema change, no reindex.

**Moved** — `BreachHistoryPanel` from always-rendered to a dedicated `Breaches` tab (with `breachesQ` search param, matching the existing tab convention).

**Added** — `BreachHistoryChart` (new file). Scatter plot above the history table:
- **X** = breach `startedAt`, **Y** = duration on log-scale (1s / 10s / 1m / 10m / 1h / 6h / 1d ticks)
- 🔴 red = past 1h grace (counts as CRITICAL seconds)
- 🟡 amber = closed inside grace (warned, not critical)
- 🟣 indigo = still ongoing
- Dotted reference line at y=3600s marks the grace boundary
- Live open-row duration uses `tradingSecondsInRange` so markers sit on the same basis as the Past-grace column and the Uptime tile
- Hover shows duration + past-grace portion + peak % of threshold

Marker density along X = frequency; marker Y = length — one view, both dimensions.

## Test plan

- [x] `pnpm test` — 1058 passing (was 1091; delta is the deleted health-score test suites)
- [x] `pnpm exec tsc --noEmit` — clean
- [x] `trunk fmt && trunk check` — clean
- [x] Browser on :3001 against prod Hasura:
  - Header no longer shows Health Score tile ✓
  - Uptime tile prominent with rollup data (USDT/USDm: 59.57% · 177 breaches) ✓
  - `Breaches` tab appears in the row, clicks through cleanly ✓
  - Chart renders 100 scatter markers on the chronically-breached pool (94 in-grace + 6 past-grace, matching the rollup ratio) ✓
  - 1h grace reference line visible ✓
  - Table indicator shows "100+ breaches" correctly ✓
- [ ] Verify on prod after merge + Vercel deploy.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

> [!NOTE]
> **Medium Risk**
> Changes breach lifecycle tracking/attribution in the indexer (anchor deferral + transition detection) and adds new paginated/filterable breach UI + uptime rollup fetching; mistakes could miscount breaches/uptime or misattribute closes, though changes are well-covered by tests and mostly additive on UI.
> 
> **Overview**
> Removes the pool-page **Health Score** metric entirely (UI components, hook, helpers, tests, and related snapshot/pool GraphQL fields), consolidating on the existing **Uptime** definition.
> 
> Fixes breach history attribution in the indexer by treating `deviationBreachStartedAt` as the source-of-truth for “in breach” and adding a `nextDeviationBreachStartedAt(..., source)` deferral for `fpmm_update_reserves` so a subsequent same-tx semantic handler (e.g. rebalance) performs the close with correct `endedByEvent`; tests were updated/added for these edge cases.
> 
> Revamps breach UX: moves breach history into a dedicated **Breaches** tab with server-side pagination/sorting, search, duration bucket + min/max filters (via new `parseDurationSeconds`), and a new Plotly scatter chart of breach start vs (log) duration. Adds an all-pools **Uptime** column computed from rollup fields fetched via a new isolated `ALL_POOLS_BREACH_ROLLUP` query, and colors uptime consistently via `uptimeColorClass`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 59dc82fe3fc30bd41ee314ab591561e6a00ad2f2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

**Comment by chapati23** [comment]
@claude review

File: ui-dashboard/src/app/pool/[poolId]/page.tsx [hunk 1]
```diff
@@ -574,6 +572,9 @@ function PoolDetail() {
             hasError={tradingLimitsError}
           />
         )}
+        {tab === "breaches" && fpmmPool && pool && (
+          <BreachHistoryPanel pool={pool} network={network} />
+        )}
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/pool/[poolId]/page.tsx:587]: Fixed in bffb734 — Breaches tab now filtered out of `visibleTabs` when `!fpmmPool`, matching the OLS pattern. Virtual pools no longer show an empty panel.

File: ui-dashboard/src/app/pool/[poolId]/page.tsx [hunk 2]
```diff
@@ -475,7 +474,6 @@ function PoolDetail() {
               ratesError={poolNeedsRates && ratesError}
             />
           </div>
-          {fpmmPool && <BreachHistoryPanel pool={pool} network={network} />}
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/app/pool/[poolId]/page.tsx:478]: Fixed in bffb734 — same fix as codex's comment above: Breaches tab filtered from `visibleTabs` when `!fpmmPool`.

File: ui-dashboard/src/lib/queries.ts [hunk 3]
```diff
@@ -423,7 +416,6 @@ export const ORACLE_SNAPSHOTS_CHART = `
       blockNumber
       txHash
       deviationRatio
-      healthBinaryValue
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/queries.ts:426]: Fixed in bffb734 — deleted both `ORACLE_SNAPSHOTS_WINDOW` and `ORACLE_SNAPSHOT_PREDECESSOR`.

File: ui-dashboard/src/components/breach-history-chart.tsx [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+              yref: "y",
+              x: 1,
+              y: grace,
+              xanchor: "right",
+              yanchor: "bottom",
+              text: "1h grace",
+              showarrow: false,
+              font: { size: 10, color: "#f59e0b" },
+            },
+          ],
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-chart.tsx:167]: Partially right — checked in-browser: shapes auto-transform for log axes (the 1h line is correctly at the "1h" tick), but annotations don't. Fixed the annotation in bffb734 by passing `Math.log10(grace)` explicitly; comment on the field explains the Plotly quirk.

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 5]
```diff
@@ -90,52 +301,165 @@ export function BreachHistoryPanel({ pool, network }: Props) {
...
+        />
+        {countCapped && (
+          <p className="px-1 pt-1 text-xs text-amber-400">
+            Showing first {ENVIO_MAX_ROWS.toLocaleString()} breaches — older
+            entries aren&apos;t visible.
+          </p>
+        )}
+        {countError && (
+          <ErrorBox message={`Count unavailable: ${countError.message}`} />
+        )}
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx:542]: Fixed in bffb734 — reworked the error flow so `countError` no longer takes down the whole panel (see reply to codex #3125513076). That makes the trailing `<ErrorBox>` reachable again (degraded-pagination case), so it stays; rewrote the message to reflect that use.

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 6]
```diff
@@ -42,37 +62,228 @@ const START_REASON_LABELS: Record<BreachEventCategory, string> = {
...
-    // (containing "type not found" / "not found in type") until the new
-    // indexer version syncs. That's a known-degraded state, not an
-    // incident — surface it in the neutral empty-state style rather than
-    // red, matching the UptimeValue tile's handling.
-    const message = error instanceof Error ? error.message : String(error);
+  }, [rows, query, getName, getTags, pool.chainId]);
+
+  // Error path collapses both count + page errors — the two share the
+  // same entity, so a schema-lag fail on one fails both.
+  if (error || countError) {
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx]: Fixed in bffb734 — `countError` now only blanks the panel when the *page* query also errors. A transient count failure degrades to a "Pagination unavailable" banner and the table keeps rendering rows.

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 7]
```diff
@@ -90,52 +355,277 @@ export function BreachHistoryPanel({ pool, network }: Props) {
...
+              maxSeconds={maxSeconds}
+              onMinCommit={onMinCommit}
+              onMaxCommit={onMaxCommit}
+            />
+          </div>
+        </div>
+
+        {isLoading && rows.length === 0 ? (
+          <p className="py-6 text-center text-sm text-slate-500">Loading…</p>
+        ) : total === 0 ? (
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx]: Fixed in bffb734 — added `countReady = countData !== undefined || countError != null` and gated the empty-state on it, so "No breaches" only fires after the count resolved.

File: ui-dashboard/src/lib/bridge-status.ts [hunk 8]
```diff
@@ -134,6 +134,65 @@ export function formatDurationShort(seconds: number): string {
...
+  let total = 0;
+  let matched = false;
+  let consumed = 0;
+  for (const m of trimmed.matchAll(pattern)) {
+    const [match, num, unit] = m;
+    const factor = unitSeconds[unit];
+    if (factor == null) return null;
+    total += Number(num) * factor;
+    matched = true;
+    consumed += match.length + (m.index ?? 0) - consumed;
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/bridge-status.ts]: Fixed in bffb734 — dead `consumed` variable removed.

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 9]
```diff
@@ -42,37 +62,282 @@ const START_REASON_LABELS: Record<BreachEventCategory, string> = {
...
+  minSeconds: number | null,
+  maxSeconds: number | null,
+): Record<string, unknown> {
+  const bucketClause = whereForBucket(bucket);
+  const rangeClauses: Record<string, unknown>[] = [];
+  if (minSeconds != null) {
+    rangeClauses.push({ durationSeconds: { _gte: String(minSeconds) } });
+  }
+  if (maxSeconds != null) {
+    rangeClauses.push({ durationSeconds: { _lte: String(maxSeconds) } });
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx]: Fixed in bffb734 — `composeWhere` now ORs the range predicate against `durationSeconds IS NULL` so ongoing breaches stay visible when a min/max filter is active. Reasoning in the docstring.

File: ui-dashboard/src/lib/queries.ts [hunk 10]
```diff
@@ -40,10 +40,9 @@ export const ALL_POOLS_WITH_HEALTH = `
       reserves0
       reserves1
       healthTotalSeconds
-      healthBinarySeconds
-      lastOracleSnapshotTimestamp
-      lastDeviationRatio
       hasHealthData
+      cumulativeCriticalSeconds
+      breachCount
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/lib/queries.ts:60]: Fixed in bffb734 — reverted the two new fields from `ALL_POOLS_WITH_HEALTH` and added a dedicated `ALL_POOLS_BREACH_ROLLUP` query wired through the existing `Promise.allSettled` in `use-all-networks-data`. Schema-lag now degrades just the Uptime column to "—".

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 11]
```diff
@@ -42,36 +62,294 @@ const START_REASON_LABELS: Record<BreachEventCategory, string> = {
...
+    limit: ENVIO_MAX_ROWS,
+  });
+  // `countReady` gates the empty state — without it, `total === 0`
+  // (defaulted from undefined) would flash a false "No breaches" card
+  // while the count query is still in flight behind the page query.
+  const countReady = countData !== undefined || countError != null;
+  const rawTotal = countData?.DeviationThresholdBreach?.length ?? 0;
+  const countCapped = rawTotal >= ENVIO_MAX_ROWS;
+  const totalPages = rawTotal > 0 ? Math.ceil(rawTotal / limit) : 1;
+  const page = Math.max(1, Math.min(rawPage, totalPages));
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx]: Fixed in 6448561 — when `countError` is set, the `page` clamp against `totalPages` is skipped and the Pagination `total` prop is computed optimistically from the returned slice length (full page → Next stays enabled, short page → Next disables). Operators can now page through older incidents while the count query is degraded.

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 12]
```diff
@@ -90,52 +374,293 @@ export function BreachHistoryPanel({ pool, network }: Props) {
...
+        {isLoading && rows.length === 0 ? (
+          <p className="py-6 text-center text-sm text-slate-500">Loading…</p>
+        ) : rows.length === 0 && countReady && rawTotal === 0 ? (
+          <EmptyBox
+            message={
+              bucket === "all"
+                ? "No deviation-threshold breaches recorded for this pool."
+                : "No breaches match this filter."
+            }
+          />
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx:451]: Fixed in 508554a — gate renamed to `countConfirmed` and requires `countData !== undefined` so a failed count never triggers the empty state. An overshoot on page 2+ combined with a `countError` now leaves the table rendering without blanking to "No breaches".

File: ui-dashboard/src/components/breach-history-panel.tsx [hunk 13]
```diff
@@ -42,36 +62,303 @@ const START_REASON_LABELS: Record<BreachEventCategory, string> = {
...
+    limit,
+    offset,
+    orderBy,
+    where,
   });
 
-  if (pool.source.includes("virtual")) return null;
+  const { data: chartData } = useGQL<{
+    DeviationThresholdBreach: DeviationThresholdBreach[];
+  }>(POOL_DEVIATION_BREACHES_ALL, { poolId: pool.id, where });
```
[the remark this answers is not in the document]
Comment (chapati23) [reply ui-dashboard/src/components/breach-history-panel.tsx]: Fixed in 59dc82f — extended `useGQL` with an optional `swrOptions` param and pass `{ revalidateOnFocus: false, revalidateOnReconnect: false }` on all three breach queries (COUNT, PAGE, ALL). Same pattern `use-bridge-gql` uses for the same rate-limit reason. The 10s refreshInterval still keeps the panel live.
